### PR TITLE
Add `no-useless-switch-case` rule

### LINKED ISSUE #1698: Rule proposal: `no-useless-switch-case`
### Description

When a switch case without a consequent is directly followed by a default case, it’s redundant. I believe it should be removed. The auto fixer can remove the redundant case.

This has also beem proposed in https://github.com/eslint/eslint/issues/14924, but it was dismissed because it doesn’t meet the criteria for new ESLint core rules.

### Fail

```js
switch (discriminant) {
  case 'redundant':
  default:
    break;
}
```

### Pass

```js
switch (discriminant) {
  case 42:
    break;
  default:
    break;
}
```

### PR DESCRIPTION
**Pull request by fisker**
Fixes #1698.

One known issue, I only checked cases where the `default` case is the last case.

Because

```js
			switch (1) {
					// This is not useless
					case 1:
					default:
							console.log('1')
					case 1:
							console.log('2')
			}
```

```js
			switch (1) {
					// This is useless
					case 1:
					default:
							console.log('1')
					case 2:
							console.log('2')
			}
```

To handle these two cases correctly, I need to compare the conditions, and it can be difficult(event impossible if the test condition is a complex expression).

Let's keep this rule simple for now.

### CODE REVIEW

File: docs/rules/no-useless-switch-case.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Disallow useless case in switch statement
+
+<!-- Do not manually modify RULE_NOTICE part. Run: `npm run generate-rule-notices` -->
+<!-- RULE_NOTICE -->
+✅ *This rule is part of the [recommended](https://github.com/sindresorhus/eslint-plugin-unicorn#recommended-config) config.*
+
+💡 *This rule provides [suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions).*
+<!-- /RULE_NOTICE -->
+
+Empty case before the last default case is useless.
```
Comment (sindresorhus) [inline docs/rules/no-useless-switch-case.md:10]:
```suggestion
An empty case before the last default case is useless.
```

File: docs/rules/no-useless-switch-case.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# Disallow useless case in switch statement
```
Comment (sindresorhus) [inline docs/rules/no-useless-switch-case.md:1]:
```suggestion
# Disallow useless case in switch statements
```

File: docs/rules/no-useless-switch-case.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		// Fallthrough
+	default:
+		handleDefaultCase();
+		break;
+}
+```
+
+```js
+switch (foo) {
+	// This is actually useless, but we only check cases where the last case is `default` case
```
Comment (sindresorhus) [inline docs/rules/no-useless-switch-case.md:58]:
```suggestion
	// This is actually useless, but we only check cases where the last case is the `default` case
```

File: rules/no-useless-switch-case.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+};
+
+const isEmptySwitchCase = node => node.consequent.every(node => isEmptyNode(node));
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => ({
+	* 'SwitchStatement[cases.length>1]'(switchStatement) {
+		const {cases} = switchStatement;
+
+		// `typescript` allow multiple `default` cases
```
Comment (sindresorhus) [inline rules/no-useless-switch-case.js:19]:
```suggestion
		// TypeScript allows multiple `default` cases
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: test/no-useless-switch-case.mjs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+				}
+				default:
+					handleDefaultCase();
+					break;
+			}
+		`,
+		outdent`
+			switch (foo) {
+				case a:
+				case (( b ))         :
```
Comment (yakov116) [inline test/no-useless-switch-case.mjs:113]:
```suggestion
				case (( b )):
```
Curious were those spaces intentional?

Comment (fisker) [reply test/no-useless-switch-case.mjs:113]: Yes

Comment (fisker) [reply test/no-useless-switch-case.mjs:113]: For this function https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1779/files#diff-d14f0af39629703585e6c7779ca7b3c2a8207c63c330bc60f2849cb40a8e14f2R16, so people can't refactor to use something like `end + 1`

File: test/no-useless-switch-case.mjs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+// 					break;
+// 				case 1:
+// 				default:
+// 					handleDefaultCase2();
+// 					break;
+// 			}
+// 		`,
+// 	],
+// 	invalid: [],
+// });
```
Comment (sindresorhus) [inline test/no-useless-switch-case.mjs:179]: If it's intentional that this is commented out, it would be good with a short comment about why.

Comment (fisker) [reply test/no-useless-switch-case.mjs:179]: It's a mistake, thanks.

**Pull request merged**